Improve loading times for web app archive pages

bourque opened the pull request:
This PR aims to improve the loading times to the web app archive pages.  Closes #676

pep8speaks commented:
Hello @bourque, Thank you for submitting the Pull Request !

* In the file [`jwql/utils/utils.py`](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/utils/utils.py):

> [Line 80:13](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/utils/utils.py#L80): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'
> [Line 432:13](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/utils/utils.py#L432): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 439:13](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/utils/utils.py#L439): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 472:21](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/utils/utils.py#L472): [E123](https://duckduckgo.com/?q=pep8%20E123) closing bracket does not match indentation of opening bracket's line
> [Line 473:21](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/utils/utils.py#L473): [E131](https://duckduckgo.com/?q=pep8%20E131) continuation line unaligned for hanging indent
> [Line 474:18](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/utils/utils.py#L474): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 503:58](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/utils/utils.py#L503): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets

* In the file [`jwql/website/apps/jwql/data_containers.py`](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/website/apps/jwql/data_containers.py):

> [Line 55:1](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/website/apps/jwql/data_containers.py#L55): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 64:1](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/website/apps/jwql/data_containers.py#L64): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 65:1](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/website/apps/jwql/data_containers.py#L65): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 497:24](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/website/apps/jwql/data_containers.py#L497): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 497:28](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/website/apps/jwql/data_containers.py#L497): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 497:38](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/website/apps/jwql/data_containers.py#L497): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 498:50](https://github.com/spacetelescope/jwql/blob/26b2f527975db9a31860abcf05f3871ef5ef78d2/jwql/website/apps/jwql/data_containers.py#L498): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

If you have not done so, please consult the [`jwql` Style Guide](https://github.com/spacetelescope/jwql/blob/master/style_guide/README.md)

bourque commented:
@tnking97 I was wondering if you would be interested and able to review this PR sometime next week (or this week, depending on when you read this)?  You are pretty familiar with how the web app pages dealing with thumbnails work, so it would be really helpful if you could look it over.  Let me know if you would have time to do that.  Thanks!

tnking97 commented:
Hi @bourque, Sure I can review this PR early next week!

bourque commented:
A few things to note about this PR:

- Putting the thumbnails in the internal storage of the dev/test/ops server already helps to speed up the web page loading times (see #694), but even with that the loading times were quite slow, especially for hefty instruments such as NIRCam.
- I determined that the major bottleneck for rendering the `archive_proposals` view was due to having to locate suitable thumbnails in the thumbnail filesystem (with `glob.glob`).  This seems unnecessary since we only really need one thumbnail to represent a given proposal, so in theory they should be easy to locate.  So to get around this, I created a new `jwql_monitor` monitor (called `generate_proposal_thumbnails.py`) that goes through the `thumbnail_filesystem`, chooses one thumbnail to represent the proposal, and copies the thumbnail to a `<proposal_id>.thumb` file.  Once that exists, the web app can now (essentially) hard-code a path to it, so no searching is necessary
- Another major bottleneck was because we were scraping the filesystem to discover files display in the web app, instead of using `astroquery.Mast` (database queries tend to be much faster than file I/O and/or pinging a network).  This is understandable, as early in the development of JWQL, `astroquery.Mast` was still being developed and was not ready to be used yet.  But I switched to using `astroquery.Mast` in this PR and it seems to be working well (and faster!)
- Another bottleneck was in the `filesytem_path` function in `utils.py`, which was checking if the file exists in the filesystem, which takes I/O & network effort.  While this is a nice check to have, it was slowing things down too much, so I created a `check_existence` parameter to optionally perform this.  For the purposes of the web app, we now skip this step.
- Another bottleneck was pinging the `utils.get_config()` function many times within various functions.  To get around this, I moved the `get_config()` file towards the beginning of the `utils` module, so that I could create a global `FILESYSTEM` variable that needs only to be defined/imported once.

bourque commented:
Awesome, thanks so much @tnking97!

tnking97 commented on jwql/jwql_monitors/generate_proposal_thumbnails.py:
```diff
@@ -0,0 +1,60 @@
+    for proposal_dir in proposal_dirs:
+        rate_thumbnails = glob.glob(os.path.join(proposal_dir, '*rate*.thumb'))
+        if rate_thumbnails:
+            thumbnail = rate_thumbnails[0]
```
Is there any case where we would expect the first thumbnail to not be representative of the set of thumbnails? This seems like a very good way to be consistent, though, and I'm not sure that it matters if it's not a representative thumbnail.

bourque replied:
Great question.  I did not re-work that logic as part of this PR, so using the first `rate` file as a thumbnail is how we currently do it, but perhaps this could be improved.  My hunch is that (at least for now), this is not a big deal, since the thumbnails are really just for aesthetics and are not intended to be used for any sort of meaningful way.

tnking97 replied:
Okay, that seems very reasonable!

tnking97 commented on jwql/utils/utils.py:
```diff
@@ -43,6 +43,109 @@
+                    "port": {"type": "string"}
+                    },
+                    "required": ['engine', 'name', 'user', 'password', 'host', 'port']
+                 },
```
It looks like there are a few little things that pep8speaks already mentioned, such as the over-indent of this bracket, but I didn't find any additional formatting issues.

tnking97 commented on jwql/utils/utils.py:
```diff
@@ -43,6 +43,109 @@
+    return settings
+
+if not ON_GITHUB_ACTIONS:
+    FILESYSTEM = get_config()['filesystem']
```
Nice way to make this run more smoothly! Minor note- I think pep8 wants two spaces before the if statement.

tnking97 commented on jwql/website/apps/jwql/data_containers.py:
```diff
@@ -1161,23 +1158,25 @@ def thumbnails_ajax(inst, proposal=None):
+    for rootname in list(data_dict['file_data'].keys()):
+        proposals.append(data_dict['file_data'][rootname]['filename_dict']['program_id'])
+        try:  # Some rootnames cannot parse out detectors
+            detectors.append(data_dict['file_data'][rootname]['filename_dict']['detector'])
```
Great, this is a good catch and should still be easy to add additional thumbnail sorting options.

tnking97 reviewed (commented):
I was excited to see these changes with lots of clever adjustments. Everything is running smoothly locally in the tests that I did. I added just a few minor comments. Did you use a profiler to figure out where the bottlenecks were?

bourque commented:
@tnking97 Thank you for the review!  I pushed some edits to address your comments, and this looks like it is good to go!
